numpy/lib/function_base.py

```diff
@@ -75,6 +76,97 @@ def iterable(y):
     return 1
 
 
+def _hist_bandwidth_estimator(a, bins):
+    """
+    A helper function to be called from histogram to deal with bandwidth estimation
+
+    if bins is a string
+        Check the corresponding estimator exists and
+        returns the optimal number of bins calculated using the estimator
+    """
+    if not isinstance(bins, basestring):
+        raise ValueError("bins = '{}', should be a string".format(bins))
+
+    def sturges(x):
+        """
+        Sturges Estimator
+        A very simplistic estimator based on the assumption of normality of the data
+        Poor performance for non-normal data, especially obvious for large X.
+        Depends only on size of the data
+        """
+        return np.ceil(np.log2(x.size)) + 1
+
+    def rice(x):
+        """
+        Rice Estimator
+        Another simple estimator, with no normality assumption.
+        It has better performance for large data, but tends to be too large
+        The number of bins is proportional to the cube root of data size (asymptotically optimal)
+        Depends only on size of the data
+        """
+        return np.ceil(2 * x.size ** (1.0 / 3))
+
+    def scott(x):
+        """
+        Scott Estimator
+        The binwidth is proportional to the standard deviation of the data and
+        inversely proportional to the cube root of data size (asymptotically optimal)
+
+        """
+        h = 3.5 * x.std() * x.size ** (-1.0 / 3)
+        if h > 0:
+            return np.ceil(x.ptp() / h)
+        return 1
+
+    def mad(data, axis=None):
+        """
+        Mean Absolute Deviation - a simple well defined function
+        Simplest implementation, even length case left to default behaviour
+        This is used by the Freedman Diaconis estimator when the IQR is 0
+        """
+        return np.median(np.absolute(data - np.median(data, axis)), axis)
+
+    def FD(x):
+        """
+        Freedman Diaconis rule using Inter Quartile Range (IQR) for binwidth
+        Considered a variation of the Scott rule with more robustness as the IQR
+        is less affected by outliers than the standard deviation.
+
+        If the IQR is 0, we return the median absolute deviation as defined above, else 1
+        Binwidth is inversely proportional to the cube root of data size (asymptotically optimal)
+        """
+        iqr = np.subtract(*np.percentile(x, [75, 25]))
+
+        if iqr == 0:
+            # unlikely, but in this case, we should replace it with another useful
+            # measure of variability - Mean Absolute Deviation (mad)
+            iqr = mad(x)
+
+        if iqr > 0:
```

Comment:
Can't really do an if-elif block. I replace the value of iqr with mad if the iqr is 0. If the mad is 0 too, then I return 1 as the number of bins, otherwise I calculate the FD estimator. 
I did move the return 1 into the `if iqr` block though